### PR TITLE
Provide the pod operation for XGBoostController 

### PR DESCRIPTION
**Pull request by merlintang**
## what is proposed in this PR
support the creating new pod operation for xgboost controller

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/xgboost-operator/12)

### CODE REVIEW

File: pkg/controller.v1alpha1/xgboost/pod.go [hunk 1]
```diff
@@ -15,16 +15,46 @@
...
 	log "github.com/sirupsen/logrus"
 	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	pylogger "github.com/kubeflow/tf-operator/pkg/logger"
+
+)
+
+const (
```
Comment (terrytangyuan) [inline pkg/controller.v1alpha1/xgboost/pod.go]: Reuse these from common: https://github.com/kubeflow/common/blob/master/job_controller/pod.go#L40-L51

File: pkg/controller.v1alpha1/xgboost/controller_test.go [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+	"k8s.io/client-go/rest"
+	"k8s.io/kubernetes/pkg/controller"
+
+	"github.com/kubeflow/tf-operator/pkg/control"
+	"github.com/kubeflow/xgboost-operator/cmd/xgboost-operator.v1alpha1/app/options"
+	jobclientset "github.com/kubeflow/xgboost-operator/pkg/client/clientset/versioned"
+	jobinformers "github.com/kubeflow/xgboost-operator/pkg/client/informers/externalversions"
+)
+
+var (
```
Comment (terrytangyuan) [inline pkg/controller.v1alpha1/xgboost/controller_test.go]: Remove unneeded vars.

File: pkg/controller.v1alpha1/xgboost/controller.go [hunk 3]
```diff
@@ -39,6 +40,25 @@ import (
...
+	// labelGroupName      = "group-name"
+	// labelXGBoostJobName = "xgboost-job-name"
+	labelXGBoostJobRole = "xgboost-job-role"
+)
+
+var (
+	// KeyFunc is the short name to DeletionHandlingMetaNamespaceKeyFunc.
+	// IndexerInformer uses a delta queue, therefore for deletes we have to use this
+	// key function but it should be just fine for non delete events.
+	KeyFunc = cache.DeletionHandlingMetaNamespaceKeyFunc
```
Comment (terrytangyuan) [inline pkg/controller.v1alpha1/xgboost/controller.go]: Reuse this from common: https://github.com/kubeflow/common/blob/48d498ab85e1ff26b75c2a3023d77b465ca25ca8/job_controller/job_controller.go#L34

File: pkg/controller.v1alpha1/xgboost/controller.go [hunk 4]
```diff
@@ -39,6 +40,25 @@ import (
 	"time"
 )
 
+const (
+	controllerName = "xgboost-operator"
+
+	// labels for pods and servers.
+	replicaTypeLabel    = "xgboost-replica-type"
+	replicaIndexLabel   = "xgboost-replica-index"
+	// labelGroupName      = "group-name"
```
Comment (terrytangyuan) [inline pkg/controller.v1alpha1/xgboost/controller.go]: Remove unused const

File: pkg/controller.v1alpha1/xgboost/pod.go [hunk 5]
```diff
@@ -15,16 +15,46 @@
...
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	common "github.com/kubeflow/common/operator/v1"
+	"github.com/kubeflow/tf-operator/pkg/common/jobcontroller"
 	"github.com/kubeflow/xgboost-operator/pkg/apis/xgboost/v1alpha1"
 	log "github.com/sirupsen/logrus"
 	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	pylogger "github.com/kubeflow/tf-operator/pkg/logger"
```
Comment (terrytangyuan) [inline pkg/controller.v1alpha1/xgboost/pod.go]: I see some imports from `tf-operator`. Could you use an alternative? I don't think we want to depend on other operators other than the common operator

File: pkg/controller.v1alpha1/xgboost/pod.go (hunk 1 above)
Comment (merlintang) [reply pkg/controller.v1alpha1/xgboost/pod.go]: it seems these constant are not exported ? I have fixed other parts based on your comments. thanks Terry.

File: pkg/controller.v1alpha1/xgboost/controller.go [hunk 6]
```diff
@@ -296,8 +310,19 @@ func (xc *XGBoostController) satisfiedExpectations(job *v1alpha1.XGBoostJob) boo
 	return false
 }
 
+// reconcilePyTorchJobs checks and updates replicas for each given PyTorchReplicaSpec.
```
Comment (terrytangyuan) [inline pkg/controller.v1alpha1/xgboost/controller.go]: pytorch -> xgboost

File: pkg/controller.v1alpha1/xgboost/controller.go [hunk 7]
```diff
@@ -296,8 +310,19 @@ func (xc *XGBoostController) satisfiedExpectations(job *v1alpha1.XGBoostJob) boo
...
+// It will requeue the job in case of an error while creating/deleting pods/services.
+func (xc *XGBoostController) reconcileXGBoostJobs(
```
Comment (terrytangyuan) [inline pkg/controller.v1alpha1/xgboost/controller.go]: Can we reuse the reconciliation function from common controller?

File: pkg/controller.v1alpha1/xgboost/pod.go [hunk 8]
```diff
@@ -37,13 +65,186 @@ func (xc *XGBoostController) DeleteService(job interface{}, name string, namespa
...
+func isNonGangSchedulerSet(xgjob *v1alpha1.XGBoostJob) bool {
+	for _, spec := range xgjob.Spec.XGBoostReplicaSpecs {
+		if spec.Template.Spec.SchedulerName != "" && spec.Template.Spec.SchedulerName != gangSchedulerName {
+			return true
+		}
+	}
+	return false
+}
+
+// GetPortFromPyTorchJob gets the port of pytorch container.
```
Comment (terrytangyuan) [inline pkg/controller.v1alpha1/xgboost/pod.go]: pytorch -> xgboost

File: pkg/controller.v1alpha1/xgboost/controller.go (hunk 7 above)
Comment (johnugeorge) [reply pkg/controller.v1alpha1/xgboost/controller.go]: Lets try to reuse common controller code for new operators.

**Comment by merlintang** [comment]
We have agreed to move on this project based on kubebuilder + common.[[issue](https://github.com/kubeflow/xgboost-operator/issues/13)] I would close this one and send other PR. thanks for your reviews and comments.

**Pull request closed without merging**